java/ql/src/semmle/code/java/frameworks/spring/SpringUtil.qll

```diff
@@ -0,0 +1,155 @@
+        "org.springframework.util.ConcurrentReferenceHashMap;Reference;false;get;;;MapKey of Argument[-1];MapKey of ReturnValue;value",
+        "org.springframework.util.ConcurrentReferenceHashMap;Reference;false;get;;;MapKey of Argument[-1];MapKey of ReturnValue;value",
+        "org.springframework.util.ConcurrentReferenceHashMap;Reference;false;get;;;MapValue of Argument[-1];MapValue of ReturnValue;value",
+        "org.springframework.util.ConcurrentReferenceHashMap;Segment;false;getReference;;;MapKey of Argument[-1];MapKey of ReturnValue;value",
+        "org.springframework.util.ConcurrentReferenceHashMap;Segment;false;getReference;;;MapValue of Argument[-1];MapValue of ReturnValue;value",
+        "org.springframework.util;ConcurrentReferenceHashMap;false;getReference;;;MapKey of Argument[-1];MapKey of ReturnValue;value",
+        "org.springframework.util;ConcurrentReferenceHashMap;false;getReference;;;MapValue of Argument[-1];MapValue of ReturnValue;value",
+        "org.springframework.util;ConcurrentReferenceHashMap;false;getSegment;;;MapKey of Argument[-1];MapKey of ReturnValue;value",
+        "org.springframework.util;ConcurrentReferenceHashMap;false;getSegment;;;MapValue of Argument[-1];MapValue of ReturnValue;value",
+        // skipped md5 digest
+        "org.springframework.util;FastByteArrayOutputStream;false;getInputStream;;;Argument[-1];ReturnValue;taint",
+        "org.springframework.util;FastByteArrayOutputStream;false;toByteArray;;;Argument[-1];ReturnValue;taint",
+        "org.springframework.util;FastByteArrayOutputStream;false;write;;;Argument[0];Argument[-1];taint",
+        "org.springframework.util;FastByteArrayOutputStream;false;writeTo;;;Argument[-1];Argument[0];taint",
+        "org.springframework.util;FileCopyUtils;false;copy;;;Argument[0];Argument[1];taint",
+        "org.springframework.util;FileCopyUtils;false;copyToByteArray;;;Argument[0];Argument[-1];taint",
+        "org.springframework.util;FileCopyUtils;false;copyToString;;;Argument[0];Argument[-1];taint",
+        "org.springframework.util;FileSystemUtils;false;copyRecursively;;;Argument[0];Argument[-1];taint",
+        "org.springframework.util;LinkedMultiValueMap;false;LinkedMultiValueMap;(Map<K,List<V>>);;MapKey of Argument[0];MapKey of ReturnValue;value",
+        "org.springframework.util;LinkedMultiValueMap;false;LinkedMultiValueMap;(Map<K,List<V>>);;Element of MapValue of Argument[0];MapValue of ReturnValue;value",
+        "org.springframework.util;LinkedMultiValueMap;false;deepCopy;;;MapKey of Argument[-1];MapValue of ReturnValue;value",
+        "org.springframework.util;LinkedMultiValueMap;false;deepCopy;;;MapValue of Argument[-1];MapValue of ReturnValue;value",
+        "org.springframework.util;MultiValueMap;true;add;;;Argument[0];MapKey of Argument[-1];value",
+        "org.springframework.util;MultiValueMap;true;add;;;Argument[1];MapValue of Argument[-1];value",
```

Comment:
Would it make sense to use `Element of MapValue` throughout? Since `MultiValueMap<K,V>` extends `Map<K,List<V>>`, it could be used as one.

Reply:
That's a good observation. Indeed that means that we should use `Element of MapValue` as you suggest.